Automattic/jetpack
# Jetpack DNA: Add full classmap support to Autoloader

This PR tweaks the Jetpack-Autoloader package to offer greater support for classmaps.

Currently, the "legacy" classmap support simply uses version `0.1` when registering the classes. That means it's totally random which class would be autoloaded should the same class exist in multiple places.

This creates classmap generation in the same way it generates for psr-4 - versioned, with `exclude-from-classmap` support

I'm testing this branch with WooCommerce + the WooCommerce REST API package which includes a classmap for it's legacy classes which are not psr-4. Example rules:

```
	'WC_REST_Tax_Classes_V1_Controller' => array(
		'version' => 'dev-test/jetpack-autoloader',
		'path'    => $vendorDir . '/woocommerce-rest-api/src/Controllers/Version1/class-wc-rest-tax-classes-v1-controller.php'
	),
	'WC_REST_Webhook_Deliveries_V1_Controller' => array(
		'version' => 'dev-test/jetpack-autoloader',
		'path'    => $vendorDir . '/woocommerce-rest-api/src/Controllers/Version1/class-wc-rest-webhook-deliveries-v1-controller.php'
	),
```

I'm not sure what process you all use to get this code into the `jetpack-autoloader` package itself which is one of the dependencies I'm using, but please let me know how I can get this shipped or if you spot any issues :)

**Testing instructions:**

This is not something you can test in the UI. You need to edit `composer.json` and add a `classmap` entry in the autoload section e.g. 

```
"autoload": {
    "classmap": [
      "includes"
    ],
    "exclude-from-classmap": [
      "includes/legacy",
      "includes/libraries"
    ]
  },
```

Then you'd need to regenerate the autoload files `composer dump-autoload` and check `vendor/composer/autoload_classmap_package.php` to see if it has the classes in your classmap definition.

**Proposed changelog entry for your changes:**

> Added classmap and exclude-from-classmap support to the jetpack-autoloader package.
